src/find_calls/find_calls.cpp

```diff
@@ -0,0 +1,200 @@
+          r.Nodes.getNodeAs<clang::DeclRefExpr>("arg");
+
+        data->call_context.insert
+          (std::pair
+          <const clang::FunctionDecl*,
+          const clang::CallExpr*>(f,c));
+
+        data->call_ref.insert
+          (std::pair
+          <const clang::CallExpr*,
+          const clang::DeclRefExpr*>(c,e));
+
+        data->call_argref.insert
+          (std::pair
+          <const clang::CallExpr*,
+          const clang::DeclRefExpr*>(c,a));
+      }
+
+  };
+
+  class AnnotateCall2
+      : public clang::ast_matchers::MatchFinder::MatchCallback {
+    private:
+      clang::CompilerInstance *ci;
+      FindCallsData *data;
+    public:
+      AnnotateCall2
+      (clang::CompilerInstance* ci,
+       FindCallsData *data)
+        :ci(ci), data(data) {}
+
+      virtual void
+      run(const clang::ast_matchers::MatchFinder::MatchResult &r)
+        override {
+
+        const clang::CallExpr *c =
+          r.Nodes.getNodeAs<clang::CallExpr>("call");
+
+        const clang::FunctionDecl *f =
+          r.Nodes.getNodeAs<clang::FunctionDecl>("context");
+
+        const clang::DeclRefExpr *e =
+          r.Nodes.getNodeAs<clang::DeclRefExpr>("ref");
+
+        const clang::StringLiteral *a =
+          r.Nodes.getNodeAs<clang::StringLiteral>("arg");
+
+        data->call_context.insert
+          (std::pair
+          <const clang::FunctionDecl*,
+          const clang::CallExpr*>(f,c));
+
+        data->call_ref.insert
+          (std::pair
+          <const clang::CallExpr*,
+          const clang::DeclRefExpr*>(c,e));
+
+        data->call_argstr.insert
+          (std::pair
+          <const clang::CallExpr*,
+          const clang::StringLiteral*>(c,a));
+      }
+
+  };
+
+  class FindCallsImpl {
+  private:
+    std::string n;
+    unsigned int a;
+    FindCallsData data;
+    clang::CompilerInstance *ci;
+
+    StatementMatcher sm1 =
+      callExpr(
+        callee(
+          implicitCastExpr(
+            has(
+              declRefExpr(
+                hasDeclaration(
+                  functionDecl(
+                    hasAnyName(
+                      n
+                    )
+                  )
+                )
+              ).bind("ref")
+            )
+          )
+        ),
+        hasArgument(a, declRefExpr().bind("arg")),
+        hasAncestor(functionDecl().bind("context"))
+      ).bind("call");
+    AnnotateCall1 cb1;
+
+    StatementMatcher sm2 =
+      callExpr(
+        callee(
+          implicitCastExpr(
+            has(
+              declRefExpr(
+                hasDeclaration(
+                  functionDecl(
+                    hasAnyName(
+                      n
+                    )
+                  )
+                )
+              ).bind("ref")
+            )
+          )
+        ),
+        hasArgument(a, stringLiteral().bind("arg")),
+        hasAncestor(functionDecl().bind("context"))
+      ).bind("call");
+    AnnotateCall2 cb2;
```

Comment:
why are there two of those?

Reply:
I think I see what you're doing. You can traverse the arguments from the call expression later, you don't even need to bind it now. That way you will even be able to go through all arguments dynamically.

Reply:
one binds a stringLiteral (aiui this will be the only supported way to later resolve a string value) and the other a reference of the argument

Reply:
Yeah, you can defer getting the argument from the call to later, it's a straight forward method call from the call expression.

Reply:
i suppose then we can just remove those binds for now and have a single matcher, do i understand you correctly?

Reply:
Correct, and then you can simply ask for the arguments in the callback by a simple method call.

Reply:
right, then the argument maps like `call_argref` can be populated by calling those on the context and reference of the function?

Reply:
Exactly.